codalab/worker/worker.py

```diff
@@ -239,12 +217,50 @@ def cleanup(self):
         logger.info("Stopped Worker. Exiting")
 
     def signal(self):
-        # When the terminate flag is False, set the stop flag to stop running
-        # the worker without changing the status of existing running bundles.
-        if not self.terminate:
+        """
+        When the pass_down_termination flag is False, set the stop flag to stop running
+        the worker without changing the status of existing running bundles. Otherwise,
+        restage all bundles that are not in the terminal states [FINISHED, RESTAGED].
+        """
+        if not self.pass_down_termination:
             self.stop = True
         else:
-            self.terminate_cli_signal = True
+            self.terminate_and_restage = True
+
+    def check_termination(self):
+        """
+        If received pass_down_termination signal from CLI to terminate the worker, wait until
+        all the existing unfinished bundles are restaged, reset runs, then stop the worker.
+        """
+        if self.terminate_and_restage:
+            if self.restage_bundles() == 0:
+                # Stop the worker
+                self.stop = True
```

Comment:
Should we rename `stop` to `terminate`?  "terminate" and "stop" are synonymous, right?  Just want to canonicalize terminology.

Reply:
That's true. I just renamed `self.stop` to `self.terminate`.